### PR TITLE
Test libgcrypt on fips mode

### PR DESCRIPTION
**Pull request by rcmadhankumar**
[CI:TOXENVS] fips

fixes: #494

### CODE REVIEW

File: tests/test_fips.py [hunk 1]
```diff
@@ -134,3 +200,101 @@ def fips_mode_setup_check(container_per_test: ContainerData) -> None:
...
+        ).stderr
+
+        assert (
+            "Hash calculation failed" in err_msg
+        ), f"Hash calculation unexpectedly succeeded for {digest}"
+
+@pytest.mark.parametrize(
+    "container_per_test", FIPS_GCRYPT_TESTER_IMAGES, indirect=True
+)
+def test_gcrypt_binary(container_per_test: ContainerData) -> None:
```
Comment (dcermak) [inline tests/test_fips.py:252]: This needs a `pytest.mark.skipif()` for sp5

**Review by dcermak — CHANGES_REQUESTED** [review]
Please remove the changes already in #561 and resolve the conflicts in `tests/test_fips.py`

**Comment by dcermak** [comment]
@rcmadhankumar please rebase your branch, the CI will not run as long as there are merge conflicts

File: tests/test_fips.py (hunk 1 above)
Comment (rcmadhankumar) [reply tests/test_fips.py]: Added skip for 15.3, 15.4, 15.5 and 15.6.
Right now libgcrypt is expected to fail on all these versions.

File: tests/test_fips.py [hunk 2]
```diff
@@ -152,3 +186,62 @@ def fips_mode_setup_check(container_per_test: ContainerData) -> None:
...
+        "crc32": "02aaf9ffb095e5e3",
+        "crc32rfc1510": "0ddebf7dcb40238c",
+        "crc24rfc2440": "35a9fbd61cc4",
+    }
+
+    for digest in FIPS_GCRYPT_DIGESTS:
+        res = container_per_test.connection.check_output(
+            f"/bin/fips-test-gcrypt {digest}"
+        )
+        assert "Digest is: " + expected_fips_gcrypt_digests.get(digest) in res
```
Comment (dcermak) [inline tests/test_fips.py:238]:
```suggestion
        assert "Digest is: " + expected_fips_gcrypt_digests[digest] in res
```

File: tests/test_fips.py [hunk 3]
```diff
@@ -152,3 +186,62 @@ def fips_mode_setup_check(container_per_test: ContainerData) -> None:
 )
 def test_openssl_fips_hashes(container_per_test: ContainerData):
     openssl_fips_hashes_test_fnct(container_per_test)
+
+
+@pytest.mark.parametrize(
+    "container_per_test", FIPS_GCRYPT_TESTER_IMAGES, indirect=True
+)
+@pytest.mark.skipif(
```
Comment (dcermak) [inline tests/test_fips.py:194]: If this is a bug that should be fixed, then the `skipif` should become an `xfail`
```suggestion
@pytest.mark.xfail(
```

**Review by dcermak — COMMENTED** [review]
Needs a rebase, otherwise LGTM

File: tests/test_fips.py [hunk 4]
```diff
@@ -236,3 +269,61 @@ def test_gnutls_binary(container_per_test: ContainerData) -> None:
...
+        "sha512_256": "04559e6cddea9a198f69bbd2a4c882d2fa3b9cda0edf18eb1582df42d88f7dba198bc813cc1b1bc065dba1bf261c1c9e3c9e92eb66c25638a447a744c38690ba",
+        "sha3-224": "7639f4678682e8c6b0968c894b9b58cac8cfbb6b44e0b7e99126229cdc0d4419c6e5dbc2e5fbb9fcb9dc6ea16d6b3cbca34aabb618e2f782",
+        "sha3-256": "9cd04d17df1c852d2b13536105c400e26d46ee0e865bd6d91d8683a2979fcb59265a271f568a62eb8c64e5cbedbdfd41d996303de25868af9b1892bda0bbcdfa",
+        "sha3-384": "b8b1ec43880fc0495a520434e379dce51fb0959f244c70b75555cae44dbac99f231bed131ee5a01ab8cd6e66e5da9b68ed7dc3dd68ec7a5898986cde448f567512a0f2cef69d1ed4e5fa2033950642c64f3a762b59576df4117cca5e7a52c188",
+        "sha3-512": "0bc583f093fb557e5aa6c3020cc93e9065a01600f3c8e1be17c5026799e4434c701a423ade5edae77491b7ee05c78cdc6e01af22bdc67fcd351fb1ab9a1936abc9a2f69add89697a8b935ce1bbc3d5c21f276b9e2377015003f98543bceffb8a2e1ec8d96fc99bd39d88ca6a0fd69d812362eda0937de4e3b3f0d94093a5b5cc",
+        "crc32": "02aaf9ffb095e5e3",
+        "crc32rfc1510": "0ddebf7dcb40238c",
+        "crc24rfc2440": "35a9fbd61cc4",
```
Comment (dcermak) [inline tests/test_fips.py:313]: CRC is cryptographically irrelevant and needn't be tested here
```suggestion
```

File: tests/files/fips-test-gcrypt.c [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        return ret_code;
+    }
+
+    if (!gcry_check_version(GCRYPT_VERSION)) {
+        fprintf(stderr, "libgcrypt version mismatch\n");
+        return ret_code;
+    }
+    gcry_control(GCRYCTL_INITIALIZATION_FINISHED, 0);
+
+    int algo = gcry_md_map_name(argv[1]);
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:31]: Nit:

```suggestion
    const int algo = gcry_md_map_name(argv[1]);
```

File: tests/files/fips-test-gcrypt.c [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <gcrypt.h>
+
+int main(int argc, char *argv[]) {
+    gcry_md_hd_t md_hd;
+    const char *mess1 = "Test Message\n";
+    const char *mess2 = "Hello World\n";
+    unsigned char md_value[128];
+    size_t md_len;
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:17]: Let's define it further below 
```suggestion
```

File: tests/files/fips-test-gcrypt.c [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    if (gcry_md_open(&md_hd, algo, GCRY_MD_FLAG_SECURE) != 0) {
+        fprintf(stderr, "Failed to create hash context\n");
+        goto cleanup;
+    }
+
+    gcry_md_write(md_hd, (const void *)mess1, strlen(mess1));
+    unsigned char hash1[64];
+    size_t hash1_size = gcry_md_get_algo_dlen(algo);
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:44]:
```suggestion
    const size_t hash1_size = gcry_md_get_algo_dlen(algo);
```

File: tests/files/fips-test-gcrypt.c [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    memcpy(hash1, gcry_md_read(md_hd, 0), hash1_size);
+
+    gcry_md_reset(md_hd);
+    gcry_md_write(md_hd, (const void *)mess2, strlen(mess2));
+    unsigned char hash2[64];
+    size_t hash2_size = gcry_md_get_algo_dlen(algo);
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:50]:
```suggestion
    const size_t hash2_size = gcry_md_get_algo_dlen(algo);
```

File: tests/files/fips-test-gcrypt.c [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    memcpy(hash2, gcry_md_read(md_hd, 0), hash2_size);
+
+    memcpy(md_value, hash1, hash1_size);
+    memcpy(md_value + hash1_size, hash2, hash2_size);
+    md_len = hash1_size + hash2_size;
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:55]:
```suggestion
    const size_t md_len = hash1_size + hash2_size;
```

File: bci_tester/fips.py [hunk 10]
```diff
@@ -64,6 +64,39 @@
...
+    "sha512",
+    "sha512_224",
+    "sha512_256",
+    "sha3-224",
+    "sha3-256",
+    "sha3-384",
+    "sha3-512",
+    "crc32",
+    "crc32rfc1510",
+    "crc24rfc2440",
```
Comment (dcermak) [inline bci_tester/fips.py:97]: These are error correction codes, they are not really interesting in a cryptographic context 
```suggestion
```

File: tests/test_fips.py [hunk 11]
```diff
@@ -65,6 +72,23 @@
             else []
         ),
     )
+   fips_gcrypt_tester_ctr = DerivedContainer(
```
Comment (dcermak) [inline tests/test_fips.py:75]: Syntax error:

```suggestion
    fips_gcrypt_tester_ctr = DerivedContainer(
```

**Review by dcermak — CHANGES_REQUESTED** [review]
Currently there's a Syntax error in test_fips.py

**Comment by rcmadhankumar** [comment]
Commit history is messed, fixing it.

File: tests/test_fips.py (hunk 2 above)
Comment (rcmadhankumar) [reply tests/test_fips.py]: fixed

File: tests/test_fips.py (hunk 4 above)
Comment (rcmadhankumar) [reply tests/test_fips.py]: removed

File: tests/files/fips-test-gcrypt.c (hunk 6 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed.

File: tests/files/fips-test-gcrypt.c (hunk 7 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed

File: tests/files/fips-test-gcrypt.c (hunk 8 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed.

File: bci_tester/fips.py (hunk 10 above)
Comment (rcmadhankumar) [reply bci_tester/fips.py]: removed

File: tests/files/fips-test-gcrypt.c (hunk 9 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed

File: tests/test_fips.py (hunk 3 above)
Comment (rcmadhankumar) [reply tests/test_fips.py]: fixed

File: tests/files/fips-test-gcrypt.c (hunk 5 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed

File: bci_tester/fips.py [hunk 12]
```diff
@@ -84,6 +84,36 @@
 #: all digests supported by gnutls
 ALL_GNUTLS_DIGESTS: Tuple[str] = NONFIPS_GNUTLS_DIGESTS + FIPS_GNUTLS_DIGESTS
 
+#: gcrypt digests that are not FIPS compliant
+NONFIPS_GCRYPT_DIGESTS: Tuple[str] = (
```
Comment (dcermak) [inline bci_tester/fips.py:88]:
```suggestion
NONFIPS_GCRYPT_DIGESTS: Tuple[str, ...] = (
```

File: bci_tester/fips.py [hunk 13]
```diff
@@ -84,6 +84,36 @@
...
+    "tiger192",
+    "md4",
+    "whirlpool",
+    "gostr3411_94",
+    "stribog256",
+    "stribog512",
+)
+
+#: FIPS compliant gcrypt digests
+FIPS_GCRYPT_DIGESTS: Tuple[str] = (
```
Comment (dcermak) [inline bci_tester/fips.py:103]:
```suggestion
FIPS_GCRYPT_DIGESTS: Tuple[str, ...] = (
```

**Review by dcermak — COMMENTED** [review]
The type hints are wrong (the issue already in the source file, so not your fault)

File: tests/files/fips-test-gcrypt.c [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+ * The hash to be used is passed as the first parameter to the binary.
+ * This program demonstrates the use of libgcrypt for computing message digests.
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <gcrypt.h>
+
+int main(int argc, char *argv[]) {
+    gcry_md_hd_t md_hd;
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:13]:
```suggestion
    gcry_md_hd_t md_hd = NULL;
```

File: tests/files/fips-test-gcrypt.c [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    const  size_t md_len = hash1_size + hash2_size;
+
+    printf("Digest is: ");
+    for (size_t i = 0; i < md_len; i++) {
+        printf("%02x", md_value[i]);
+    }
+    ret_code = 0;
+
+cleanup:
+    gcry_md_close(md_hd);
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:63]: If you don't check for `NULL`, you'll get a segfault if the message digest is unknown (branch in line 31).
```suggestion
    if (md_hd != NULL) {
        gcry_md_close(md_hd);
    }
```

File: tests/files/fips-test-gcrypt.c [hunk 16]
```diff
@@ -0,0 +1,65 @@
+/** Simple C program that will calculate the hash of a fixed string using
+ * libgcrypt.
+ *
+ * The hash to be used is passed as the first parameter to the binary.
+ * This program demonstrates the use of libgcrypt for computing message digests.
+ */
+#include <stdio.h>
+#include <stdlib.h>
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:8]: Not needed:
```suggestion
```

File: tests/files/fips-test-gcrypt.c [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+        fprintf(stderr, "libgcrypt version mismatch\n");
+        return ret_code;
+    }
+    gcry_control(GCRYCTL_INITIALIZATION_FINISHED, 0);
+
+    const int algo = gcry_md_map_name(argv[1]);
+    if (algo == 0) {
+        fprintf(stderr, "Unknown message digest %s\n", argv[1]);
+        goto cleanup;
+    }
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:34]: You can also return here immediately, as you haven't touched `md_hd` yet
```suggestion
    if (algo == 0) {
        fprintf(stderr, "Unknown message digest %s\n", argv[1]);
        return ret_code;
    }
```

File: tests/files/fips-test-gcrypt.c [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+    gcry_md_write(md_hd, (const void *)mess1, strlen(mess1));
+    unsigned char hash1[64];
+    const size_t hash1_size = gcry_md_get_algo_dlen(algo);
+    memcpy(hash1, gcry_md_read(md_hd, 0), hash1_size);
+
+    gcry_md_reset(md_hd);
+    gcry_md_write(md_hd, (const void *)mess2, strlen(mess2));
+    unsigned char hash2[64];
+    const  size_t hash2_size = gcry_md_get_algo_dlen(algo);
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:49]:
```suggestion
    const size_t hash2_size = gcry_md_get_algo_dlen(algo);
```

File: tests/files/fips-test-gcrypt.c [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    memcpy(hash2, gcry_md_read(md_hd, 0), hash2_size);
+
+    memcpy(md_value, hash1, hash1_size);
+    memcpy(md_value + hash1_size, hash2, hash2_size);
+    const  size_t md_len = hash1_size + hash2_size;
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:54]:
```suggestion
    const size_t md_len = hash1_size + hash2_size;
```

File: tests/test_fips.py [hunk 20]
```diff
@@ -236,3 +269,65 @@ def test_gnutls_binary(container_per_test: ContainerData) -> None:
...
+            f"/bin/fips-test-gcrypt {digest}"
+        )
+        assert "Digest is: " + expected_fips_gcrypt_digests[digest] in res
+
+    for digest in NONFIPS_GCRYPT_DIGESTS:
+        if OS_VERSION == "15.6" and digest == "md5":
+            # md5 mistakenly calculates the digest
+            # bugzilla: 1229903
+            # skipping md5 on OS Version 15.6
+            continue
```
Comment (dcermak) [inline tests/test_fips.py:325]: Is this any different from the xfail in the function decorator?

Comment (rcmadhankumar) [reply tests/test_fips.py:325]: There is not much difference. Only thing is, when we do this way the test case passes on 15.6 and we are just making sure here that except md5 all the other NON FIPS algorithms works as expected.

Comment (rcmadhankumar) [reply tests/test_fips.py:325]: I just wanted this to be more accuarete, but we can use xfail as well.

File: tests/files/fips-test-gcrypt.c (hunk 14 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed

File: tests/files/fips-test-gcrypt.c (hunk 15 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed

File: tests/files/fips-test-gcrypt.c (hunk 16 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed.

File: tests/files/fips-test-gcrypt.c (hunk 17 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed

File: tests/files/fips-test-gcrypt.c (hunk 18 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed.

File: tests/files/fips-test-gcrypt.c (hunk 19 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gcrypt.c]: fixed.

File: tests/test_fips.py (hunk 20 above)
Comment (rcmadhankumar) [reply tests/test_fips.py]: Changed to xfail.

File: tests/files/fips-test-gcrypt.c [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+    if (!gcry_check_version(GCRYPT_VERSION)) {
+        fprintf(stderr, "libgcrypt version mismatch\n");
+        return ret_code;
+    }
+    gcry_control(GCRYCTL_INITIALIZATION_FINISHED, 0);
+
+    const int algo = gcry_md_map_name(argv[1]);
+    if (algo == 0) {
+        fprintf(stderr, "Unknown message digest %s\n", argv[1]);
+        return ret_code
```
Comment (dcermak) [inline tests/files/fips-test-gcrypt.c:32]: syntax error:
```suggestion
        return ret_code;
```

File: tests/test_fips.py [hunk 22]
```diff
@@ -237,3 +270,95 @@ def test_gnutls_binary(container_per_test: ContainerData) -> None:
...
+
+        if (
+            OS_VERSION == "15.3"
+            and digest in unsupported_non_fips_gcrypt_digests_15_3
+        ):
+            assert (
+                non_fips_call.rc == 1
+                and f"Unknown message digest {digest}" in non_fips_call.stderr
+            ), f"Hash calculation unexpectedly supported for {digest} in os version {OS_VERSION}"
+            continue
```
Comment (dcermak) [inline tests/test_fips.py:359]: Imho it would be better to not call `/bin/fips-test-grypt` on the unsupported hashes.

File: tests/test_fips.py [hunk 23]
```diff
@@ -237,3 +270,95 @@ def test_gnutls_binary(container_per_test: ContainerData) -> None:
...
+
+        assert (
+            res.rc == 0
+            and "Digest is: " + expected_fips_gcrypt_digests[digest]
+            in res.stdout
+        )
+
+    for digest in NONFIPS_GCRYPT_DIGESTS:
```
Comment (dcermak) [inline tests/test_fips.py:335]: Either you can exclude the unsupported hashes here:
```suggestion
    for digest in set(NONFIPS_GCRYPT_DIGESTS) - (set(unsupported_non_fips_gcrypt_digests_15_3) if OS_VERSION=="15.3" else {}):
```

File: bci_tester/fips.py [hunk 24]
```diff
@@ -86,6 +86,36 @@
...
+    "tiger",
+    "tiger2",
+    "tiger192",
+    "md4",
+    "whirlpool",
+    "gostr3411_94",
+    "stribog256",
+    "stribog512",
+    "md5",
+)
```
Comment (dcermak) [inline bci_tester/fips.py:102]: Or you can add it here (imho simpler):
```suggestion
NONFIPS_GCRYPT_DIGESTS: Tuple[str, ...] = (
    "ripemd160",
    "tiger",
    "tiger2",
    "tiger192",
    "md4",
    "whirlpool",
    "gostr3411_94",
    "stribog256",
    "stribog512",
    "md5",
)

if OS_VERSION != "15.3":
    NONFIPS_GCRYPT_DIGESTS += ("sm3",)
```

File: tests/test_fips.py [hunk 25]
```diff
@@ -237,3 +270,95 @@ def test_gnutls_binary(container_per_test: ContainerData) -> None:
...
+
+        if (
+            OS_VERSION == "15.3"
+            and digest in unsupported_fips_gcrypt_digests_15_3
+        ):
+            assert (
+                res.rc == 1
+                and f"Unknown message digest {digest}" in res.stderr
+            ), f"Hash calculation unexpectedly supported for {digest} in os version {OS_VERSION}"
+            continue
```
Comment (dcermak) [inline tests/test_fips.py:327]: Same as with the non-fips hash algorithms: should be imho never executed

**Review by dcermak — APPROVED** [review]

**Pull request merged**